File: 프로그래머스/1/178871. 달리기 경주/달리기 경주.js
function solution(players, callings) {
    // 이름 -> 현재 등수 (index) 맵
    const nameToIndex = new Map();
    for (let i = 0; i < players.length; i++) {
        nameToIndex.set(players[i], i);
    }

    for (const name of callings) {
        const idx = nameToIndex.get(name);
        const prevPlayer = players[idx - 1];

        // swap players[idx] <-> players[idx - 1]
        [players[idx], players[idx - 1]] = [players[idx - 1], players[idx]];

        // update nameToIndex map
        nameToIndex.set(name, idx - 1);
        nameToIndex.set(prevPlayer, idx);
    }

    return players;
}

File: 프로그래머스/1/178871. 달리기 경주/달리기 경주.kt
class Solution {
    fun solution(players: Array<String>, callings: Array<String>): Array<String> {
        val nameToIndex = HashMap<String, Int>()
        
        // 선수 이름 -> 인덱스 매핑
        for (i in players.indices) {
            nameToIndex[players[i]] = i
        }

        for (name in callings) {
            val idx = nameToIndex[name]!!
            val prevIdx = idx - 1
            val prevPlayer = players[prevIdx]

            // 선수 위치 스왑
            players[prevIdx] = name
            players[idx] = prevPlayer

            // 인덱스 업데이트
            nameToIndex[name] = prevIdx
            nameToIndex[prevPlayer] = idx
        }

        return players
    }
}
